base-server.d.ts

```diff
@@ -0,0 +1,704 @@
+ * class MyLoguxHack extends BaseServer {
+ *   …
+ * }
+ * ```
+ */
+export class BaseServer {
+  constructor(opts: LoguxBaseServerOptions)
+
+  /**
+   * Server options.
+   *
+   * ```js
+   * console.log('Server options', server.options.subprotocol)
+   * ```
+   */
+  options: LoguxBaseServerOptions
+  reporter: LoguxBaseServerOptions['reporter'] | (() => void)
+
+  /**
+   * Production or development mode.
+   *
+   * ```js
+   * if (server.env === 'development') {
+   *   logDebugData()
+   * }
+   * ```
+   */
+  env: Required<LoguxBaseServerOptions['env']>
+
+  /**
+   * Server unique ID.
+   *
+   * ```js
+   * console.log('Error was raised on ' + server.nodeId)
+   * ```
+   */
+  nodeId: string
+
+  /**
+   * Server actions log.
+   *
+   * ```js
+   * server.log.each(finder)
+   * ```
+   */
+  log: Log
+
+  /**
+   * Connected clients.
+   *
+   * ```js
+   * for (let i in server.connected) {
+   *   console.log(server.connected[i].remoteAddress)
+   * }
+   * ```
+   */
+  connected: {
+    [key: string]: ServerClient
+  }
+
+  /**
+   * Set authenticate function. It will receive client credentials
+   * and node ID. It should return a Promise with `true` or `false`.
+   *
+   * @param authenticator The authentication callback.
+   *
+   * ```js
+   * server.auth(async (userId, token) => {
+   *   const user = await findUserByToken(token)
+   *   return !!user && userId === user.id
+   * })
+   * ```
+   */
+  auth<Credentials = any>(authenticator: LoguxAuthenticator<Credentials>): void
+
+  /**
+   * Start WebSocket server and listen for clients.
+   *
+   * @returns When the server has been bound.
+   */
+  listen(): Promise<void>
+
+  /**
+   * Subscribe for synchronization events. It implements nanoevents API.
+   * Supported events:
+   *
+   * * `error`: server error during action processing.
+   * * `fatal`: server error during loading.
+   * * `clientError`: wrong client behaviour.
+   * * `connected`: new client was connected.
+   * * `disconnected`: client was disconnected.
+   * * `preadd`: action is going to be added to the log.
+   *   The best place to set `reasons`.
+   * * `add`: action was added to the log.
+   * * `clean`: action was cleaned from the log.
+   * * `processed`: action processing was finished.
+   * * `subscribed`: channel initial data was loaded.
+   *
+   * @param event The event name.
+   * @param listener The listener function.
+   * @returns Unbind listener from event.
+   *
+   * ```js
+   * server.on('error', error => {
+   *   trackError(error)
+   * })
+   * ```
+   */
+  on(event: 'fatal' | 'clientError', listener: (err: Error) => void): void
+  on<Action extends LoguxBaseAction = LoguxBaseAction>(
+    event: 'error',
+    listener: (err: Error, action: Action, meta: LoguxMeta) => void
+  ): void
+  on(
+    event: 'connected' | 'disconnected',
+    listener: (server: ServerClient) => void
+  ): void
+  on<Action extends LoguxBaseAction = LoguxBaseAction>(
```

Comment:
`LoguxBaseAction` works only for `logux/subscribe`, `logux/processed` and `logux/undo`.

User's action can be with any keys. The only limit is to have `type: string`.